daemon/api.go

```diff
@@ -967,6 +976,119 @@ func deleteCapability(c *Command, r *http.Request) Response {
 	}
 }
 
+// skillGrant is a part of response from getSkills().
+type skillGrant struct {
+	Snap string `json:"snap"`
+	Name string `json:"name"`
+}
+
+// skillInfo is a part of response from getSkills().
+type skillInfo struct {
+	Snap  string       `json:"snap"`
+	Name  string       `json:"name"`
+	Type  string       `json:"type"`
+	Label string       `json:"label"`
+	Slots []skillGrant `json:"slots"`
+}
+
+// getSkills returns a list of all the skills and which slots use them.
+func getSkills(c *Command, r *http.Request) Response {
+	var skills []skillInfo
+	for _, skill := range c.d.skills.AllSkills("") {
+		var slots []skillGrant
+		for _, slot := range c.d.skills.GrantsOf(skill.Snap, skill.Name) {
+			slots = append(slots, skillGrant{
+				Snap: slot.Snap,
+				Name: slot.Name,
+			})
+		}
+		skills = append(skills, skillInfo{
+			Snap:  skill.Snap,
+			Name:  skill.Name,
+			Type:  skill.Type,
+			Label: skill.Label,
+			Slots: slots,
+		})
+	}
+	return SyncResponse(skills)
+}
+
+// SkillAction is an action performed on the skill system.
+type SkillAction struct {
+	Action string       `json:"action"`
+	Skill  skills.Skill `json:"skill,omitempty"`
+	Slot   skills.Slot  `json:"slot,omitempty"`
+}
+
+// changeSkills controls the skill system.
+// Skills can be granted to and revoked from slots.
+// When enableInternalSkillActions is true skills and slots can also be
+// explicitly added and removed.
+func changeSkills(c *Command, r *http.Request) Response {
+	var a SkillAction
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&a); err != nil || a.Action == "" {
```

Comment:
The `a.Action == ""` should have its own error report saying the action is missing. It's painful to interact with remote systems that do not tell what's actually wrong with the request.